Refine chain definition, add test cases

quanlinc opened the pull request:
<!--- # Please remember to use `describe` and `it`in the tests! see https://github.com/flow-typed/flow-typed/blob/master/CONTRIBUTING.md for details. --->

- Links to documentation: https://ramdajs.com/docs/#chain
- Link to GitHub or NPM: https://github.com/ramda/ramda
- Type of contribution: refactor

Other notes:
Restrict input list to be read only, add type coverage for the case when second argument is a function

jedwards1211 committed:
fix(@apollo/*): export ApolloError type (#3613)

dependabot[bot] committed:
Bump mixin-deep from 1.3.1 to 1.3.2 in /cli (#3626)

Bumps [mixin-deep](https://github.com/jonschlinkert/mixin-deep) from 1.3.1 to 1.3.2.
- [Release notes](https://github.com/jonschlinkert/mixin-deep/releases)
- [Commits](https://github.com/jonschlinkert/mixin-deep/compare/1.3.1...1.3.2)

Signed-off-by: dependabot[bot] <support@github.com>

LoganBarnett committed:
fix ramda complement (#3634)

Ramda's complement returns a function whose type is the same as the input
function provided. The runtime difference is that the boolean value is flipped
when the function is called. Flow doesn't handle expressing this very well, and
thus doesn't preserve the returned function's type. Until we have some mechanism
to do this, we can write a bunch of arity variants - or generate them, as was
done here. The libdef and its accompanying test are both generated.

If you want to generate higher arity to support or want to revisit the
generation, the original generation was done here:
https://github.com/LoganBarnett/typedef-gen

LoganBarnett committed:
fix qs 6.5.x and add qs 6.9.x (#3631)

* fix qs 6.5.x and add qs 6.9.x

The qs libdefs used the only Function, Object, and sometimes any types - some of
which cause issues in newer versions of Flow and/or flow-typed. The old qs 6.5.x
tests have had the bare minimum done to get the tests passing again. Most of the
work was done against a new (copied) set of libdefs for 6.9.x. I did not know
what versions added which arguments, or if certain arguments were omitted on the
original 6.5.x, and perusing the documentation for qs did not reveal this in a
timely manner. It seemed to make sense make a clean break to 6.9.x. The 6.9.x
version of qs should be entirely backwards compatible with 6.5.x, since only
new, optional arguments have been added.

* fix qs linting issues re: implicit inexact objects

migueloller committed:
Fix typo in styled-components_v4.x.x (#3609)

supersonicclay committed:
[long_v4.x.x] Add definitions (#3642)

Commit:
Refine chain definition, add test cases

gyrfalcon commented on definitions/npm/ramda_v0.26.x/flow_v0.104.x-/ramda_v0.26.x.js:
```diff
@@ -652,8 +652,11 @@ declare module ramda {
 
-  declare function chain<A, B>(f: (x: A) => B[], xs: A[]): B[]
-  declare function chain<A, B>(f: (x: A) => B[]): (xs: A[]) => B[]
+  declare function chain<A, B>(f: (x: A) => B[], xs: $ReadOnlyArray<A>): B[];
```
For consistency, I'd suggest changing `B[]` to `Array<B>` (to parallel using `$ReadOnlyArray<A>`).

quanlinc replied:
Done

Commit:
Use 'Array' instead of '[]'

Commit:
Missed one place

Allaoua9 committed:
[@reach/router_v1.1.x] Add state to DomLocation and Redirect props (#3647)

gantoine committed:
Update list of maintainers

gyrfalcon commented:
Looks good to me

madsnedergaard committed:
[antd_v3.x.x] Updates enum for "type" in Button (to support new "link" type) (#3658)

lukeapage committed:
[lodash_v4.x.x] Use $ReadOnlyArray for `min` and `max` (#3661)

Commit:
[dompurify_v1.x.x] Remove deprecated $SubType and fix hooks (#3660)

serhalp committed:
[jest_24.x.x] Add jest-extended toIncludeSameMembers (#3663)

See https://github.com/jest-community/jest-extended#toincludesamemembersmembers.

LoganBarnett commented on definitions/npm/ramda_v0.26.x/flow_v0.104.x-/test_ramda_v0.26.x_list.js:
```diff
@@ -822,4 +824,30 @@ const str: string = "hello world";
+  describe('chain', () => {
+    const readOnlyArray = ['1', '2', '3'];
+
+    it('should only accept read only list as input', () => {
```
"should only" probably should not have "only". We want to support both cases, right?

quanlinc replied:
Yeah, this one is done earlier before we realize we want read only in but mutable out. I'll have the wording fixed.

quanlinc replied:
Done

LoganBarnett commented on definitions/npm/ramda_v0.26.x/flow_v0.104.x-/test_ramda_v0.26.x_list.js:
```diff
@@ -822,4 +824,30 @@ const str: string = "hello world";
   }))([1, 2, 3])(["1", "2", "3"]);
+
+  describe('chain', () => {
+    const readOnlyArray = ['1', '2', '3'];
```
I'd roll these out into the individual tests so the tests are not coupled with each other.

quanlinc replied:
By applying type restriction to the variable, I found 'last' definition not taking read only in causing the test to fail, I'll save the day to fix the family of 'head, last, tail' and etc.

quanlinc replied:
Done

LoganBarnett commented on definitions/npm/ramda_v0.26.x/flow_v0.104.x-/test_ramda_v0.26.x_list.js:
```diff
@@ -822,4 +824,30 @@ const str: string = "hello world";
+      const result2: Array<string> = chain(duplicate)(readOnlyArray);
+    });
+
+    it('should error out when type casting', () => {
```
I think this test could be better described as "The output type is an array of the input function's return element type". We could workshop that one to be better too ;) But basically it's good to capture the overall behavior you want to see or the intent of the test and the types working in conjunction.

quanlinc replied:
I've changed to the one I used in other PRs, let me know if you like some special care on this one to phrase the way you proposed :)

quanlinc replied:
Done

LoganBarnett commented on definitions/npm/ramda_v0.26.x/flow_v0.76.x-v0.103.x/ramda_v0.26.x.js:
```diff
@@ -617,8 +617,11 @@ declare module ramda {
+  declare function chain<A, B>(f: (x: A) => Array<B>, xs: $ReadOnlyArray<A>): Array<B>
+  declare function chain<A, B>(f: (x: A) => Array<B>): (xs: $ReadOnlyArray<A>) => Array<B>
+
+  //if second argument is a function, chain(f, g)(x) is equivalent to f(g(x), x)
```
I like this context! I'd expand upon it (since it feels very non-sequitur here) by including the surrounding documentation from Ramda:

```
// Additional chain functionality documented at https://ramdajs.com/docs/#chain:
//
// Dispatches to the chain method of the second argument, if present, according to the FantasyLand Chain spec (https://github.com/fantasyland/fantasy-land#chain).
//
// If second argument is a function, chain(f, g)(x) is equivalent to f(g(x), x).
//
// Acts as a transducer if a transformer is given in list position.
```

quanlinc replied:
Great! I'll shamelessly copy your comment :)

quanlinc replied:
Done

LoganBarnett reviewed (changes requested):
Left some comments.

klaesra committed:
Use readonly for lodash4 pickBy and omitBy (#3648)

wilt00 committed:
[haversine_v1.1.x] ArrayCoordinates as 2-tuple (#3662)

julienw committed:
Add libdefs for dotenv v8 (#3664)

quanlinc committed:
change input of uniq to be read only array (#1) (#3659)

* change input of uniq to be read only array

julienw committed:
Add the libdef for @koa/cors (#3668)

julienw committed:
Add libdefs for koa-json-body (#3670)

vonovak committed:
fix: remove invalid react-navigation-stack back button props (#3669)

Commit:
refactor some test cases, additional comment on definition

lukeapage committed:
[lodash_v4.x.x] xorBy accepts $ReadOnlyArray (#3673)

mischnic committed:
[semver_6.2.x] Add definitions (#3674)

julienw committed:
[content-type_v1.x.x] Add definitions (#3675)

julienw committed:
[supertest_v4.x.x] Add the 'called' property (#3676)

quanlinc committed:
Merge branch 'master' into ramda-chain

quanlinc commented:
PR to flow-typed: 
https://github.com/flow-typed/flow-typed/pull/3680/

quanlinc committed:
[ramda_v0.26.x] Intersection readonly refactor (#3682)

Commit:
[lodash_v4.x.x] take input object as `$ReadOnly` (#3683)

quanlinc committed:
[axios_v0.19.x] New options definition (#3681)

eawtf committed:
[axios_v0.19.x] Split Flow version to align with recent changes. (#3689) (#3691)

Commit:
docs(ignoreDeps): fix `bundled` typo (#3690)

Commit:
[query-string_v6.x.x] Add 'comma' value to array format (#3688)

valscion committed:
Update react-dnd_v2.x.x.js to avoid $Supertype (#3645)

* Trigger test run for react-dnd_v2 test

* Remove broken $ExpectError tests for react-dnd_v2

See https://travis-ci.org/flow-typed/flow-typed/builds/613489940 for a
test failure, showing how these tests were faulty already

* Update react-dnd_v2.x.x.js to avoid $Supertype

While upgrading flow through 0.98, which turns on deprecated-utility by default I came across this use of the Supertype utility.

Substituting in $Shape seems to work well

* Make react-dnd most recent types target from v0.98.x forwards

* Remove support of Flow v0.98.x for react-dnd types

* Revert "Remove support of Flow v0.98.x for react-dnd types"

This reverts commit 900a288ec342e2a094f2e68faf0686b95f763826.

It's better to support Flow v0.98.x with a deprecation warning than not
support it at all. CI does not check such an old version of Flow, so it
won't cause CI failures.

Co-authored-by: Andy McVitty <Andy.mcvitty@gmail.com>

quanlinc committed:
Allow 'adjust' to accept read only input array (#3671)

* make adjust to work with read only input array

* Add missed import

* Add missing definition for flow prior to 0.104

quanlinc committed:
Refactor 'nth' to take read only array (#3665)

* Refactor 'nth' to take read only array

* refactor test cases, change nth definition to be exact type

* refine tests and fix a typo

* refactor tests

* Fix a typo

* fix typo and refactor test

* Fix typo

Co-authored-by: Andrew Smith <andrewsmith@alumni.stanford.edu>

AndrewSouthpaw committed:
Merge branch 'master' into ramda-chain
